src/balancer-strategy-manager/BalancerV2WeightedPoolStrategyManager.sol

```diff
@@ -0,0 +1,226 @@
+
+    IAsset[] memory assets = new IAsset[](TOKEN_COUNT);
+    for (uint256 i = 0; i < TOKEN_COUNT; ) {
+      IERC20(tokenConfig[i].token).safeIncreaseAllowance(address(VAULT), _tokenAmounts[i]);
+      assets[i] = IAsset(tokenConfig[i].token);
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    userData = abi.encode(
+      WeightedPoolUserData.JoinKind.EXACT_TOKENS_IN_FOR_BPT_OUT,
+      _tokenAmounts,
+      0
+    );
+
+    IBalancerVault.JoinPoolRequest memory request = IBalancerVault.JoinPoolRequest({
+      assets: assets,
+      maxAmountsIn: _tokenAmounts,
+      userData: userData,
+      fromInternalBalance: false
+    });
+
+    VAULT.joinPool(POOL_ID, address(this), address(this), request);
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function withdraw(uint256 bpt) external onlyOwnerOrGuardian returns (uint256[] memory) {
+    return _withdraw(bpt);
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function emergencyWithdraw() external onlyWithdrawable returns (uint256[] memory) {
+    uint256 bptAmount = IERC20(address(POOL)).balanceOf(address(this));
+
+    return _withdraw(bptAmount);
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function getTokenConfig(uint256 id) external view override returns (TokenConfig memory) {
+    return tokenConfig[id];
+  }
+
+  /// @dev withdraws token from pool
+  function _withdraw(uint256 bptAmount) internal returns (uint256[] memory tokenAmounts) {
+    // IERC20(address(POOL)).safeIncreaseAllowance(address(VAULT), bptAmount);
+
+    uint256[] memory minAmountsOut = new uint256[](TOKEN_COUNT);
+    IAsset[] memory assets = new IAsset[](TOKEN_COUNT);
+    for (uint256 i = 0; i < TOKEN_COUNT; ) {
+      assets[i] = IAsset(tokenConfig[i].token);
+      minAmountsOut[i] = 0;
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    bytes memory userData = abi.encode(
+      WeightedPoolUserData.ExitKind.EXACT_BPT_IN_FOR_TOKENS_OUT,
+      bptAmount
+    );
+    IBalancerVault.ExitPoolRequest memory request = IBalancerVault.ExitPoolRequest({
+      assets: assets,
+      minAmountsOut: minAmountsOut,
+      userData: userData,
+      toInternalBalance: false
+    });
+
+    VAULT.exitPool(POOL_ID, address(this), payable(address(this)), request);
+
+    tokenAmounts = _backTokens();
+  }
+
+  /// @dev send back remaining tokens to provider
+  function _backTokens() internal returns (uint256[] memory) {
+    uint256[] memory tokenAmounts = new uint256[](TOKEN_COUNT);
+
+    for (uint256 i = 0; i < TOKEN_COUNT; ) {
+      tokenAmounts[i] = _backToken(IERC20(tokenConfig[i].token), tokenConfig[i].provider);
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    return tokenAmounts;
+  }
+
+  /// @dev send back remaining tokens to provider
+  function _backToken(IERC20 token, address provider) internal returns (uint256 tokenAmount) {
+    tokenAmount = token.balanceOf(address(this));
+
+    if (tokenAmount > 0) {
+      token.approve(address(POOL), 0);
```

Comment:
This won't work with USDT, you need to `forceApprove()`

More generally, why are we using `safeIncrease` somewhere and elsewhere `approve()`?